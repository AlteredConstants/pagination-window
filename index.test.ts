import getPaginationWindow from "./index";

function getConfig({ offset = 0, limit = 10, total = 50 }) {
  return { offset, limit, total };
}

describe("getPaginationWindow", () => {
  it("should return null if no config object is provided", () => {
    const windows = {
      nullInput: getPaginationWindow(<any>null),
      undefinedInput: getPaginationWindow(<any>undefined),
    };
    expect(windows).toEqual({ nullInput: null, undefinedInput: null });
  });

  it("should return null if config is not an object", () => {
    const window = getPaginationWindow(<any>"ðŸ’¥");
    expect(window).toBeNull();
  });

  it("should return null if offset, limit, or total are not numbers", () => {
    const windows = {
      offset: getPaginationWindow(<any>{ offset: "ðŸ’¥", limit: 10, total: 50 }),
      limit: getPaginationWindow(<any>{ offset: 0, limit: "ðŸ’¥", total: 50 }),
      total: getPaginationWindow(<any>{ offset: 0, limit: 10, total: "ðŸ’¥" }),
    };
    expect(windows).toEqual({ offset: null, limit: null, total: null });
  });

  it("should return null if offset, limit, or total are negative", () => {
    const windows = {
      offset: getPaginationWindow(<any>{ offset: -1, limit: 10, total: 50 }),
      limit: getPaginationWindow(<any>{ offset: 0, limit: -1, total: 50 }),
      total: getPaginationWindow(<any>{ offset: 0, limit: 10, total: -1 }),
    };
    expect(windows).toEqual({ offset: null, limit: null, total: null });
  });

  it("should return null if limit is zero", () => {
    const window = getPaginationWindow({ offset: 0, limit: 0, total: 50 });
    expect(window).toBeNull();
  });

  it("should create a window with 1 page if total is zero", () => {
    const window = getPaginationWindow({ offset: 0, limit: 10, total: 0 });
    expect(window).toMatchSnapshot();
  });

  it("should create a window without ellipses", () => {
    const window = getPaginationWindow({ offset: 40, limit: 10, total: 50 });
    expect(window).toMatchSnapshot();
  });

  it("should create a window with ellipses", () => {
    const window = getPaginationWindow({ offset: 40, limit: 10, total: 90 });
    expect(window).toMatchSnapshot();
  });

  it("should create only one page if total is zero", () => {
    const window = getPaginationWindow({ offset: 0, limit: 10, total: 0 });
    expect(window).toContainEqual(
      expect.objectContaining({ type: "page", number: 1 }),
    );
    expect(window).not.toContainEqual(
      expect.objectContaining({ type: "page", number: 2 }),
    );
  });

  it("should create only one page if total is less than the limit", () => {
    const window = getPaginationWindow({ offset: 0, limit: 10, total: 1 });
    expect(window).toContainEqual(
      expect.objectContaining({ type: "page", number: 1 }),
    );
    expect(window).not.toContainEqual(
      expect.objectContaining({ type: "page", number: 2 }),
    );
  });

  it("should not create a new page if total is on a limit boundary", () => {
    const window = getPaginationWindow({ offset: 0, limit: 10, total: 10 });
    expect(window).toContainEqual(
      expect.objectContaining({ type: "page", number: 1 }),
    );
    expect(window).not.toContainEqual(
      expect.objectContaining({ type: "page", number: 2 }),
    );
  });

  it("should create two pages if total is twice the limit", () => {
    const window = getPaginationWindow({ offset: 0, limit: 10, total: 20 });
    expect(window).toContainEqual(
      expect.objectContaining({ type: "page", number: 2 }),
    );
  });

  it("should disable 'previous' nav if offset is in the first page", () => {
    const window = getPaginationWindow({ offset: 0, limit: 10, total: 50 });
    expect(window[0]).toMatchObject({
      type: "navigation",
      direction: "previous",
      isDisabled: true,
    });
  });

  it("should disable 'next' nav if offset is in the last page", () => {
    const window = getPaginationWindow({ offset: 50, limit: 10, total: 50 });
    expect(window[window.length - 1]).toMatchObject({
      type: "navigation",
      direction: "next",
      isDisabled: true,
    });
  });

  it("should display only back ellipsis if offset is in the first half of the window", () => {
    const window = getPaginationWindow({ offset: 30, limit: 10, total: 90 });
    expect(window[2]).toMatchObject({ type: "page" });
    expect(window[window.length - 3]).toMatchObject({ type: "ellipsis" });
  });

  it("should display only front ellipsis if offset is in the second half of the window", () => {
    const window = getPaginationWindow({ offset: 60, limit: 10, total: 90 });
    expect(window[2]).toMatchObject({ type: "ellipsis" });
    expect(window[window.length - 3]).toMatchObject({ type: "page" });
  });

  it("should display both ellipses if offset is in the middle of the window", () => {
    const window = getPaginationWindow({ offset: 40, limit: 10, total: 90 });
    expect(window[2]).toMatchObject({ type: "ellipsis" });
    expect(window[window.length - 3]).toMatchObject({ type: "ellipsis" });
  });
});
