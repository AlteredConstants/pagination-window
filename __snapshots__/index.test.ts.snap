// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPaginationWindow should create a window with 1 page if total is zero 1`] = `
Array [
  Object {
    "direction": "previous",
    "isCurrent": false,
    "isDisabled": true,
    "key": "navigation-previous",
    "number": 1,
    "offset": 0,
    "type": "navigation",
  },
  Object {
    "isCurrent": true,
    "isDisabled": false,
    "key": "page-1",
    "number": 1,
    "offset": 0,
    "type": "page",
  },
  Object {
    "direction": "next",
    "isCurrent": false,
    "isDisabled": true,
    "key": "navigation-next",
    "number": 1,
    "offset": 0,
    "type": "navigation",
  },
]
`;

exports[`getPaginationWindow should create a window with ellipses 1`] = `
Array [
  Object {
    "direction": "previous",
    "isCurrent": false,
    "isDisabled": false,
    "key": "navigation-previous",
    "number": 4,
    "offset": 30,
    "type": "navigation",
  },
  Object {
    "isCurrent": false,
    "isDisabled": false,
    "key": "page-1",
    "number": 1,
    "offset": 0,
    "type": "page",
  },
  Object {
    "isCurrent": false,
    "isDisabled": true,
    "key": "ellipsis-front",
    "type": "ellipsis",
  },
  Object {
    "isCurrent": false,
    "isDisabled": false,
    "key": "page-3",
    "number": 3,
    "offset": 20,
    "type": "page",
  },
  Object {
    "isCurrent": false,
    "isDisabled": false,
    "key": "page-4",
    "number": 4,
    "offset": 30,
    "type": "page",
  },
  Object {
    "isCurrent": true,
    "isDisabled": false,
    "key": "page-5",
    "number": 5,
    "offset": 40,
    "type": "page",
  },
  Object {
    "isCurrent": false,
    "isDisabled": false,
    "key": "page-6",
    "number": 6,
    "offset": 50,
    "type": "page",
  },
  Object {
    "isCurrent": false,
    "isDisabled": false,
    "key": "page-7",
    "number": 7,
    "offset": 60,
    "type": "page",
  },
  Object {
    "isCurrent": false,
    "isDisabled": true,
    "key": "ellipsis-back",
    "type": "ellipsis",
  },
  Object {
    "isCurrent": false,
    "isDisabled": false,
    "key": "page-9",
    "number": 9,
    "offset": 80,
    "type": "page",
  },
  Object {
    "direction": "next",
    "isCurrent": false,
    "isDisabled": false,
    "key": "navigation-next",
    "number": 6,
    "offset": 50,
    "type": "navigation",
  },
]
`;

exports[`getPaginationWindow should create a window without ellipses 1`] = `
Array [
  Object {
    "direction": "previous",
    "isCurrent": false,
    "isDisabled": false,
    "key": "navigation-previous",
    "number": 4,
    "offset": 30,
    "type": "navigation",
  },
  Object {
    "isCurrent": false,
    "isDisabled": false,
    "key": "page-1",
    "number": 1,
    "offset": 0,
    "type": "page",
  },
  Object {
    "isCurrent": false,
    "isDisabled": false,
    "key": "page-2",
    "number": 2,
    "offset": 10,
    "type": "page",
  },
  Object {
    "isCurrent": false,
    "isDisabled": false,
    "key": "page-3",
    "number": 3,
    "offset": 20,
    "type": "page",
  },
  Object {
    "isCurrent": false,
    "isDisabled": false,
    "key": "page-4",
    "number": 4,
    "offset": 30,
    "type": "page",
  },
  Object {
    "isCurrent": true,
    "isDisabled": false,
    "key": "page-5",
    "number": 5,
    "offset": 40,
    "type": "page",
  },
  Object {
    "direction": "next",
    "isCurrent": false,
    "isDisabled": true,
    "key": "navigation-next",
    "number": 5,
    "offset": 40,
    "type": "navigation",
  },
]
`;
